components/board/pi/impl/board.go

```diff
@@ -190,28 +228,139 @@ func (pi *piPigpio) performConfiguration(
 		ar := &board.MCP3008AnalogReader{channel, bus, ac.ChipSelect}
 		pi.analogs[ac.Name] = board.SmoothAnalogReader(ar, ac, pi.logger)
 	}
+	return nil
+}
 
-	// setup interrupts
-	pi.interrupts = map[string]board.DigitalInterrupt{}
-	pi.interruptsHW = map[uint]board.DigitalInterrupt{}
-	for _, c := range cfg.DigitalInterrupts {
-		bcom, have := broadcomPinFromHardwareLabel(c.Pin)
-		if !have {
-			return errors.Errorf("no hw mapping for %s", c.Pin)
+// This is a helper function for digital interrupt reconfiguration. It finds the key in the map
+// whose value is the given interrupt, and returns that key and whether we successfully found it.
+func findInterruptName(
+	interrupt board.ReconfigurableDigitalInterrupt,
+	interrupts map[string]board.ReconfigurableDigitalInterrupt,
+) (string, bool) {
+	for key, value := range interrupts {
+		if value == interrupt {
+			return key, true
+		}
+	}
+	return "", false
+}
+
+// This is a very similar helper function, which does the same thing but for broadcom addresses.
+func findInterruptBcom(
+	interrupt board.ReconfigurableDigitalInterrupt,
+	interruptsHW map[uint]board.ReconfigurableDigitalInterrupt,
+) (uint, bool) {
+	for key, value := range interruptsHW {
+		if value == interrupt {
+			return key, true
 		}
+	}
+	return 0, false
+}
+
+func (pi *piPigpio) reconfigureInterrupts(ctx context.Context, cfg *genericlinux.Config) error {
+	// We reuse the old interrupts when possible.
+	oldInterrupts := pi.interrupts
+	oldInterruptsHW := pi.interruptsHW
+	newInterrupts := map[string]board.ReconfigurableDigitalInterrupt{}
+	newInterruptsHW := map[uint]board.ReconfigurableDigitalInterrupt{}
 
-		di, err := board.CreateDigitalInterrupt(c)
+	// This begins as a set of all interrupts, but we'll remove the ones we reuse. Then, we'll
+	// close whatever is left over.
+	interruptsToClose := make(map[board.ReconfigurableDigitalInterrupt]struct{}, len(oldInterrupts))
+	for _, interrupt := range oldInterrupts {
```

Comment:
[q] COuld `oldInterrupts` and `oldInterruptsHW` ever have a mismatch, and would that be missed by line 334 onwards?

Reply:
I don't believe so: any time one is modified, the other is, too (both as `pi.interrupts` and `pi.interruptsHW`, and as `oldInterrupts` and `oldInterruptsHW`). I'll put in a comment about that to make it clearer (and double-check that this really is an invariant we have! It's supposed to be).